core/amber/src/main/scala/edu/uci/ics/texera/web/resource/WorkflowWebsocketResource.scala

```diff
@@ -158,34 +190,247 @@ class WorkflowWebsocketResource {
+      cacheSinkOperators = sessionCacheSinkOperators(session.getId)
+    }
+    var operatorRecord: mutable.HashMap[String, WorkflowVertex] = null
+    if (!sessionOperatorRecord.contains(session.getId)) {
+      operatorRecord = mutable.HashMap[String, WorkflowVertex]()
+    } else {
+      operatorRecord = sessionOperatorRecord(session.getId)
+    }
+
+    val workflowInfo = WorkflowInfo(request.operators, request.links, request.breakpoints)
+    workflowInfo.cachedOperatorIDs = request.cachedOperatorIDs
+    logger.info("Cached operators: {}.", cachedOperators.toString())
+    logger.info("request.cachedOperatorIDs: {}.", request.cachedOperatorIDs)
+    val workflowRewriter = new WorkflowRewriter(
+      workflowInfo,
+      cachedOperators.clone(),
+      cacheSourceOperators.clone(),
+      cacheSinkOperators.clone(),
+      operatorRecord.clone(),
+      opResultStorage
+    )
+
+    val invalidSet = workflowRewriter.cacheStatusUpdate()
+
+    val cacheStatusMap = request.cachedOperatorIDs
+      .filter(cachedOperators.contains)
+      .map(id => {
+        if (cachedOperators.contains(id)) {
+          if (!invalidSet.contains(id)) {
+            (id, CacheStatus.CACHE_VALID)
+          } else {
+            (id, CacheStatus.CACHE_INVALID)
+          }
+        } else {
+          (id, CacheStatus.CACHE_INVALID)
+        }
+      })
+      .toMap
+
+    val cacheStatusUpdateEvent = CacheStatusUpdateEvent(cacheStatusMap)
+    send(session, cacheStatusUpdateEvent)
+  }
+
+  val sessionCachedOperators: mutable.HashMap[String, mutable.HashMap[String, OperatorDescriptor]] =
+    mutable.HashMap[String, mutable.HashMap[String, OperatorDescriptor]]()
+  val sessionCacheSourceOperators
+      : mutable.HashMap[String, mutable.HashMap[String, CacheSourceOpDesc]] =
+    mutable.HashMap[String, mutable.HashMap[String, CacheSourceOpDesc]]()
+  val sessionCacheSinkOperators: mutable.HashMap[String, mutable.HashMap[String, CacheSinkOpDesc]] =
+    mutable.HashMap[String, mutable.HashMap[String, CacheSinkOpDesc]]()
+  val sessionOperatorRecord: mutable.HashMap[String, mutable.HashMap[String, WorkflowVertex]] =
+    mutable.HashMap[String, mutable.HashMap[String, WorkflowVertex]]()
+
+  val opResultStorageConfig = ConfigFactory.load("application")
+  val opResultSwitch = opResultStorageConfig.getString("cache.switch").equals("on")
+  var opResultStorage: OpResultStorage = _
+  if (opResultSwitch) {
+    if (Constants.storage.equals("memory")) {
+      opResultStorage = new MemoryOpResultStorage()
+      logger.info("use memory storage for materialization")
+    } else if (Constants.storage.equals("JCS")) {
+      opResultStorage = new JCSOpResultStorage()
+      logger.info("use JCS for materialization")
+    } else if (Constants.storage.equals("mongodb")) {
+      logger.info("use mongodb for materialization")
+      opResultStorage = new MongoOpResultStorage()
+    } else {
+      logger.info("invalid storage config")
+      System.exit(-1)
```

Comment:
don't log info and exit -1, just throw an Exception.

Reply:
fixed